src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java

```diff
@@ -8,108 +8,110 @@
 		public final long eventTimeMS;
 
-		InflightMessage(Position p, long eventTimeMS) {
+		InflightMessage(Position p, long eventTimeMS, long messageID) {
 			this.position = p;
 			this.isComplete = false;
 			this.sendTimeMS = System.currentTimeMillis();
 			this.eventTimeMS = eventTimeMS;
+			this.messageID = messageID;
 		}
 
 		long timeSinceSendMS() {
 			return System.currentTimeMillis() - sendTimeMS;
 		}
 	}
 
-	private static final long INIT_CAPACITY = 1000;
-	private static final double COMPLETE_PERCENTAGE_THRESHOLD = 0.9;
+	// number of total messages we allow to be outstanding at once
+	private static final int DEFAULT_CAPACITY = 10000;
+
+	// number of messages we allow to be processed while the head of the queue is stuck
+	private static final int DEFAULT_STUCK_HEAD_DETECTION = 3000;
+
+	// how long before we consider the head of the queue stuck
+	private final long producerAckTimeoutMS;
+
+	private final int stuckHeadDetection;
 
 	private final LinkedHashMap<Position, InflightMessage> linkedMap;
 	private final MaxwellContext context;
-	private final long capacity;
-	private final long producerAckTimeoutMS;
-	private final double completePercentageThreshold;
-	private volatile boolean isFull;
+	private final Semaphore semaphore;
+	private long messageCount = 0;
 
 	public InflightMessageList(MaxwellContext context) {
-		this(context, INIT_CAPACITY, COMPLETE_PERCENTAGE_THRESHOLD);
+		this(context, DEFAULT_CAPACITY, DEFAULT_STUCK_HEAD_DETECTION);
 	}
 
-	public InflightMessageList(MaxwellContext context, long capacity, double completePercentageThreshold) {
+	public InflightMessageList(MaxwellContext context, int capacity, int stuckHeadDetection) {
 		this.context = context;
 		this.producerAckTimeoutMS = context.getConfig().producerAckTimeout;
-		this.completePercentageThreshold = completePercentageThreshold;
+		this.stuckHeadDetection = stuckHeadDetection;
 		this.linkedMap = new LinkedHashMap<>();
-		this.capacity = capacity;
+		this.semaphore = new Semaphore(capacity);
 	}
 
-	public void addMessage(Position p, long eventTimestampMillis) throws InterruptedException {
-		synchronized (this.linkedMap) {
-			while (isFull) {
-				this.linkedMap.wait();
-			}
-
-			InflightMessage m = new InflightMessage(p, eventTimestampMillis);
-			this.linkedMap.put(p, m);
+	public long waitForSlot() throws InterruptedException {
+		this.semaphore.acquire();
+		return ++this.messageCount;
+	}
 
-			if (linkedMap.size() >= capacity) {
-				isFull = true;
+	public void freeSlot(long messaageID) {
+		// If the head is stuck for the length of time (configurable) and majority of the messages have completed,
+		// we assume the head will unlikely get acknowledged, hence terminate Maxwell.
+		// This gatekeeper is the last resort since if anything goes wrong,
+		// producer should have raised exceptions earlier than this point when all below conditions are met.
+		if (producerAckTimeoutMS > 0) {
+			Iterator<InflightMessage> it = iterator();
```

Comment:
I don't think you're allowed to do this without synchronizing on something:

https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html
>  If multiple threads access a linked hash map concurrently, and at least one of the threads modifies the map structurally, it must be synchronized externally.

It may be that the inconsistencies we'd see by not synchronizing are benign (e.g. outdated results), but it may also be that we'd hit NullPointerExceptions, we can't know for sure.

It seems like we only need the head of the inflight list for this code. If we don't want to synchronize here, we could instead store a `volatile` field referencing the head message whenever we mutate the underlying list, then this code would only need to read that field.